File: frontend/src/types/task.ts
import type { UserSummary } from '@/types/artifact'

export type TaskStatus = 'OPEN' | 'IN_PROGRESS' | 'REVIEWING' | 'COMPLETED' | 'CLOSED'

export type ParticipantStatus = 'ACTIVE' | 'LEFT'

export type SubmissionStatus = 'SUBMITTED' | 'REVISION_REQUIRED' | 'ACCEPTED' | 'REJECTED'

export interface TaskListItem {
  id: number
  title: string
  creator: UserSummary
  status: TaskStatus
  deadline_at?: string | null
  completed_at?: string | null
  closed_at?: string | null
  created_at: string
  updated_at: string
}

export interface TaskParticipant {
  id: number
  task_id: number
  user: UserSummary
  status: ParticipantStatus
  joined_at: string
  left_at?: string | null
}

export interface TaskRead extends TaskListItem {
  description: string
  my_participation?: TaskParticipant | null
  participant_count: number
  submission_count: number
}

export interface TaskListResponse {
  items: TaskListItem[]
  page: number
  page_size: number
  total: number
}

export interface TaskParticipantListResponse {
  items: TaskParticipant[]
}

export interface SubmissionArtifactSummary {
  id: number
  title: string
  status: string
}

export interface SubmissionTaskSummary {
  id: number
  title: string
  status: TaskStatus
}

export interface TaskSubmission {
  id: number
  task_id: number
  participant_id: number
  participant: UserSummary
  artifact: SubmissionArtifactSummary
  round_no: number
  note?: string | null
  status: SubmissionStatus
  is_current: boolean
  submitted_at: string
  revision_requested_at?: string | null
  decided_at?: string | null
  decider?: UserSummary | null
  decision_note?: string | null
  task?: SubmissionTaskSummary | null
}

export interface TaskSubmissionListResponse {
  items: TaskSubmission[]
  page: number
  page_size: number
  total: number
}

export interface ArtifactTaskSourceListResponse {
  items: TaskSubmission[]
}

export interface TaskInput {
  title: string
  description: string
  deadline_at: string | null
}

export interface TaskSubmissionInput {
  artifact_id: number
  note?: string | null
}

File: backend/app/schemas/task.py
"""API schemas for tasks."""

from __future__ import annotations

from datetime import UTC, datetime
from decimal import Decimal
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator

from app.schemas.artifact import UserSummary
from app.schemas.common import UtcJsonModel


class TaskStatus(StrEnum):
    OPEN = "OPEN"
    IN_PROGRESS = "IN_PROGRESS"
    REVIEWING = "REVIEWING"
    COMPLETED = "COMPLETED"
    CLOSED = "CLOSED"


TERMINAL_TASK_STATUSES = (TaskStatus.COMPLETED, TaskStatus.CLOSED)


class ParticipantStatus(StrEnum):
    ACTIVE = "ACTIVE"
    LEFT = "LEFT"


class SubmissionStatus(StrEnum):
    SUBMITTED = "SUBMITTED"
    REVISION_REQUIRED = "REVISION_REQUIRED"
    ACCEPTED = "ACCEPTED"
    REJECTED = "REJECTED"


class TaskCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=100_000)
    deadline_at: datetime | None = None

    @field_validator("title", "description")
    @classmethod
    def strip_required_text(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("must not be blank")
        return value

    @field_validator("deadline_at")
    @classmethod
    def normalize_deadline(cls, value: datetime | None) -> datetime | None:
        if value is None:
            return None
        if value.tzinfo is None:
            raise ValueError("must include a UTC offset")
        return value.astimezone(UTC)


class TaskUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, min_length=1, max_length=100_000)
    deadline_at: datetime | None = None

    @field_validator("title", "description")
    @classmethod
    def strip_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        if not value:
            raise ValueError("must not be blank")
        return value

    @field_validator("deadline_at")
    @classmethod
    def normalize_deadline(cls, value: datetime | None) -> datetime | None:
        if value is None:
            return None
        if value.tzinfo is None:
            raise ValueError("must include a UTC offset")
        return value.astimezone(UTC)


class TaskListItem(UtcJsonModel):
    id: int
    title: str
    creator: UserSummary
    status: TaskStatus
    competition_id: int | None = None
    competition_title: str | None = None
    deadline_at: datetime | None
    completed_at: datetime | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class TaskParticipantRead(UtcJsonModel):
    id: int
    task_id: int
    user: UserSummary
    status: ParticipantStatus
    joined_at: datetime
    left_at: datetime | None


class TaskRead(TaskListItem):
    description: str
    my_participation: TaskParticipantRead | None = None
    participant_count: int = 0
    submission_count: int = 0


class TaskListResponse(BaseModel):
    items: list[TaskListItem]
    page: int
    page_size: int
    total: int


class TaskParticipantListResponse(BaseModel):
    items: list[TaskParticipantRead]


class SubmissionArtifactSummary(BaseModel):
    id: int
    title: str
    status: str


class TaskCompetitionSummary(BaseModel):
    id: int
    title: str
    lifecycle_status: str


class SubmissionTaskSummary(BaseModel):
    id: int
    title: str
    status: TaskStatus
    competition: TaskCompetitionSummary | None = None


class SubmissionReviewSummary(UtcJsonModel):
    raw_score: Decimal
    comment: str | None
    reviewed_at: datetime
    reviewer: UserSummary

    @field_serializer("raw_score", when_used="json")
    @classmethod
    def serialize_raw_score(cls, value: Decimal) -> str:
        return f"{value:.2f}"


class TaskSubmissionRead(UtcJsonModel):
    id: int
    task_id: int
    participant_id: int
    participant: UserSummary
    artifact: SubmissionArtifactSummary
    round_no: int
    note: str | None
    status: SubmissionStatus
    is_current: bool
    submitted_at: datetime
    revision_requested_at: datetime | None
    decided_at: datetime | None
    decider: UserSummary | None = None
    decision_note: str | None
    task: SubmissionTaskSummary | None = None
    competition_review: SubmissionReviewSummary | None = None
    competition_rank: int | None = None
    competition_award: str | None = None


class TaskSubmissionListResponse(BaseModel):
    items: list[TaskSubmissionRead]
    page: int
    page_size: int
    total: int


class ArtifactTaskSourceListResponse(BaseModel):
    items: list[TaskSubmissionRead]


class TaskSubmissionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    artifact_id: int
    note: str | None = Field(default=None, max_length=2_000)

    @field_validator("note")
    @classmethod
    def strip_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None


def _require_nonblank_note(value: str) -> str:
    value = value.strip()
    if not value:
        raise ValueError("must not be blank")
    return value


class SubmissionRevisionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    note: str = Field(min_length=1, max_length=2_000)

    @field_validator("note")
    @classmethod
    def strip_note(cls, value: str) -> str:
        return _require_nonblank_note(value)


class SubmissionAcceptRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    note: str | None = Field(default=None, max_length=2_000)

    @field_validator("note")
    @classmethod
    def strip_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None


class SubmissionRejectRequest(SubmissionRevisionRequest):
    pass
